### PR TITLE
Add data transforms for flux form

### PR DESCRIPTION
**Pull request by oliverwm1**
Add data transforms which implement the flux-form fine-res approach. This includes following transforms:

1. Qm_flux and downward sfc radiation from Qm
2. Q2_flux and precipitation from Q2
3. Qm from Qm_flux and downward sfc radiation
4. Q2 from Q2_flux and precipitation
5. downward sfc radiation from Qm and upward surface fluxes
6. precipitation from Q2 and E

Added public API:
- new transforms for use before/after ML prediction

- [x] Tests added

### CODE REVIEW

File: external/vcm/vcm/data_transform.py [hunk 1]
```diff
@@ -77,6 +97,205 @@ def Q2_from_dQ2_pQ2(ds):
...
+    for derivation of transform."""
+    Qm_flux = -vcm.mass_cumsum(ds["Qm"], ds[DELP], dim="z")
+    toa_rad_flux = ds[DSW_TOA] - ds[USW_TOA] - ds[ULW_TOA]
+    upward_surface_mse_flux = ds[LHF] + ds[SHF] + ds[USW_SFC] + ds[ULW_SFC]
+
+    # add TOA boundary condition
+    Qm_flux = Qm_flux.pad(z=(1, 0), constant_values=0.0)
+    Qm_flux += toa_rad_flux
+    if include_temperature_nudging:
+        Qm_flux += ds[COL_T_NUDGE]
```
Comment (AnnaKwa) [inline external/vcm/vcm/data_transform.py:131]: Is this a 3d field where the value at level z is the cumulative sum up to z?

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
The new set of transform functions are pretty complex relative to the other data transforms, and share similar computations- could you move the bulk of the calculations outside of `data_transforms` and then just import and call them in the registered functions? `data_transforms` seems like it will get a lot of use / more functions added in the future, so it would nice to try keep the transforms concise.

That would also make it easier to have tests for the functions. Writing a test case for them might be a bit annoying but I think since these outputs are central to the flux form experiments it would be reassuring to have tests for the flux-to-tendency and tendency-to-flux functions.

File: external/vcm/vcm/data_transform.py (hunk 1 above)
Comment (oliverwm1) [reply external/vcm/vcm/data_transform.py]: Qm_flux is 3D, `ds[COL_T_NUDGE]` is 2D (it is the column-integrated fine-res nudging tendency, which is an output of the fine-res loader).

File: external/vcm/tests/test_calc.py [hunk 2]
```diff
@@ -166,3 +166,17 @@ def test_round_trip_mse_temperature_tendency():
     Qm = vcm.moist_static_energy_tendency(Q1, Q2)
     round_tripped_Q1 = vcm.temperature_tendency(Qm, Q2)
     xr.testing.assert_allclose(Q1, round_tripped_Q1)
+
+
+def test_mass_cumsum():
+    da = xr.DataArray(np.reshape(np.arange(0, 40, 2), (5, 4)), dims=["x", "z"])
+    delp = xr.DataArray(np.reshape(np.arange(0, 40, 2), (5, 4)), dims=["x", "z"])
+    vcm.mass_cumsum(da, delp, dim="z")
+
```
[the remark this answers is not in the document]
Comment (oliverwm1) [reply external/vcm/tests/test_calc.py:175]: I was just testing that the function doesn't raise an error, but yeah I can check shape.

File: external/vcm/vcm/data_transform.py [hunk 3]
```diff
@@ -77,6 +97,205 @@ def Q2_from_dQ2_pQ2(ds):
...
+        DLW_SFC,
+        DSW_SFC,
+        DSW_TOA,
+        ULW_SFC,
+        ULW_TOA,
+        USW_SFC,
+        USW_TOA,
+        LHF,
+        SHF,
+        COL_T_NUDGE,
```
[the remark this answers is not in the document]
Comment (oliverwm1) [reply external/vcm/vcm/data_transform.py:117]: Good question. Typically this is called on data that is output from the fine-res loader, in which case the `COL_T_NUDGE` will be available. If we were to use this in a prognostic run setting, that might not be the case. At this point that's not a required use case, but if it becomes one, I guess we could add a separate transform that doesn't require `COL_T_NUDGE`?

**Comment by oliverwm1** [comment]
> That would also make it easier to have tests for the functions.

I did have tests of the relevant data transforms that ensured tendency -> flux -> tendency would round trip successfully. But agreed it will be useful to have some functions/tests that are a bit more separate from the data transform infrastructure.

**Review by AnnaKwa — APPROVED** [review]
Thanks Oli! I guess since there are quite a few inputs for the flux to/from tendency calculations, the overall transform function length didn't change much with the requested refactor, but I think it is nice that both Q2/Qm transforms use common functions that can be tested.

**Pull request merged**